utils/vlog/yosys/json.py

```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/env python3
+"""
+This is intended to provide a range of helper functions around the output of
+Yosys' `write_json`. Depending on the tasks, this may need to be flattened
+and/or in AIG format. In any case, at minimum `proc` and usually `prep` should
+be used before outputting the JSON.
+"""
+
+import os, sys
+import json
+
+class YosysModule:
+    def __init__(self, name, module_data):
+        self.name = name
+        self.data = module_data
+
+    @property
+    def ports(self):
+        """List of ports on a module.
+
+        Returns a list of tuples:
+        -------
+        name : str
+        width : int
+            The width in bits
+        dir : str
+            The direction, should be either `input` or `output`
+        """
+        plist = []
+        for port, pdata in self.data["ports"].items():
+            plist.append((port, len(pdata["bits"]), pdata["direction"]))
+        return plist
+
+    @property
+    def cells(self):
+        """List of cells of a module, excluding Yosys-internal cells
+        beginning with $.
+
+        Returns a list of tuples:
+        -------
+        name : str
+        type: str
+        """
+        clist = []
+        for cell, cdata in self.data["cells"].items():
+            if cell.startswith("$"):
+                continue
+            clist.append((cell, cdata["type"]))
+        return clist
+
+    def cell_type(self, cell):
+        """Return the type of a given cell"""
+        for cname, cdata in self.data["cells"].items():
+            if cname == cell:
+                return cdata["type"]
+        return None
+
+    @property
+    def module_attrs(self):
+        """All attributes of a module as a dictionary"""
+        return self.data["attributes"]
+
+    def attr(self, attr, defval = None):
```

Comment:
Maybe use `__getattr__` ? https://blog.rmotr.com/python-magic-methods-and-getattr-75cf896b3f88